File: apps/server/api/src/routes/assist.ts
/**
 * Desktop Assistant Support Routes.
 *   GET  /api/assist/:sessionId/commands / GET /api/assist/commands — list skills for slash autocomplete
 *   GET  /api/assist/:sessionId/search   / GET /api/assist/search   — FFF-backed fuzzy file search (@ file refs)
 * Both resolve the session's working directory as the search root.
 */
import { Hono, type Context } from "hono";
import { discoverSkills } from "@/agent/src/systemprompt/discover-skills.js";
import { getSharedSessionStorage } from "@/agent/src/session/storage.js";
import { searchFiles } from "@/api/src/services/assist.service.js";
import type { SlashCommandInfo } from "@console/types";

export const assistRoutes = new Hono();
const sessionStorage = getSharedSessionStorage();

async function handleCommands(c: Context) {
  const sessionId = c.req.param("sessionId") ?? c.req.query("sessionId") ?? "";
  const session = sessionId ? sessionStorage.loadSession(sessionId) : null;
  const cwd = session?.header.cwd ?? process.cwd();

  const skills = await discoverSkills({ cwd });

  const commands: SlashCommandInfo[] = skills
    .filter((s) => !s.hide)
    .map((skill) => ({
      name: skill.name,
      description: skill.description ?? "",
      builtin: false,
    }));

  return c.json({ success: true, data: commands });
}

async function handleSearch(c: Context) {
  const sessionId = c.req.param("sessionId") ?? c.req.query("sessionId") ?? "";
  const query = (c.req.query("q") ?? c.req.query("query") ?? "").trim();
  const session = sessionId ? sessionStorage.loadSession(sessionId) : null;
  const root = c.req.query("root") ?? session?.header.cwd ?? process.cwd();
  // Opt-out for file-only consumers (e.g. desktop ⌘P): defaults to true so
  // the @-mention picker keeps directory results.
  const includeDirs =
    c.req.query("includeDirs") !== "false" && c.req.query("includeDirs") !== "0";

  try {
    // When query is empty (user just typed "@") fall back to a broad scan so
    // the desktop autocomplete can show something instead of a blank popup.
    const items = await searchFiles(root, query || ".", undefined, includeDirs);
    return c.json({ success: true, data: { root, query, items } });
  } catch (err) {
    const errorMsg = err instanceof Error ? err.message : String(err);
    return c.json({ success: false, error: errorMsg }, 400);
  }
}

assistRoutes.get("/assist/:sessionId/commands", handleCommands);
assistRoutes.get("/assist/commands", handleCommands);
assistRoutes.get("/assist/:sessionId/search", handleSearch);
assistRoutes.get("/assist/search", handleSearch);
assistRoutes.get("/assist/files", handleSearch);

File: apps/server-go/internal/routes/assist.go
// Desktop assistant support routes (/api/assist/*). Port of
// apps/server/api/src/routes/assist.ts: slash-command autocomplete (init +
// discovered skills) and fff-backed @-mention file search scoped to the
// session cwd.
package routes

import (
	"os"
	"strings"

	"github.com/gofiber/fiber/v2"

	"github.com/Adelodunpeter25/console/apps/server-go/internal/services"
	"github.com/Adelodunpeter25/console/apps/server-go/internal/types"
)

const (
	initCommandName        = "init"
	initCommandDescription = "Generate a console.toml with project run scripts for the Run tab"
)

func registerAssistRoutes(app *fiber.App, sessions *services.SessionService, fs *services.FsService, skills *services.SkillsService) {
	handleCommands := func(c *fiber.Ctx) error {
		cwd := resolveSessionCwd(c, sessions)
		commands := []types.SlashCommandInfo{
			{Name: initCommandName, Description: initCommandDescription, Builtin: true},
		}
		for _, skill := range skills.Discover(cwd) {
			commands = append(commands, types.SlashCommandInfo{
				Name: skill.Name, Description: skill.Description, Builtin: false,
			})
		}
		return c.JSON(fiber.Map{"success": true, "data": commands})
	}

	handleSearch := func(c *fiber.Ctx) error {
		query := strings.TrimSpace(c.Query("q"))
		if query == "" {
			query = strings.TrimSpace(c.Query("query"))
		}
		root := c.Query("root")
		if root == "" {
			root = resolveSessionCwd(c, sessions)
		}
		includeDirs := c.Query("includeDirs") != "false" && c.Query("includeDirs") != "0"
		searchQuery := query
		if searchQuery == "" {
			// Empty query (user just typed "@") falls back to a broad scan.
			searchQuery = "."
		}
		items, err := fs.SearchFiles(root, searchQuery, 50, includeDirs)
		if err != nil {
			return fail400(c, err)
		}
		// Ensure we always return a valid response even if no items found
		if items == nil {
			items = []types.FileSearchResult{}
		}
		return c.JSON(fiber.Map{"success": true, "data": fiber.Map{
			"root": root, "query": query, "items": items,
		}})
	}

	app.Get("/api/assist/:sessionId/commands", handleCommands)
	app.Get("/api/assist/commands", handleCommands)
	app.Get("/api/assist/:sessionId/search", handleSearch)
	app.Get("/api/assist/search", handleSearch)
	app.Get("/api/assist/files", handleSearch)
}

// resolveSessionCwd mirrors the TS handler: explicit session id param or
// query, else the server cwd.
func resolveSessionCwd(c *fiber.Ctx, sessions *services.SessionService) string {
	sessionID := c.Params("sessionId")
	if sessionID == "" {
		sessionID = c.Query("sessionId")
	}
	if sessionID != "" {
		if loaded, err := sessions.Load(sessionID, 0, 0); err == nil && loaded != nil && loaded.Header.Cwd != "" {
			if _, err := os.Stat(loaded.Header.Cwd); err == nil {
				return loaded.Header.Cwd
			}
		}
	}
	cwd, err := os.Getwd()
	if err != nil {
		return "."
	}
	return cwd
}
